### DISCUSSION TITLE
Great difference in results of equation of states between dp.predict and lammps.minimize

### DISCUSSION THREAD

**Discussion by PolyuWeldingSpock** [deepmodeling/deepmd-kit#2980] [Q&A]
I have trained a deep potential with the help of DPGEN. To validate the accuracy of deep potential, I have calculated equation of states by scaling lattice constant from 0.90 to 1.10. I used both dp.predict function and minimize command in lammps. However, these two methods gives me total different results.

First is dp.predict result, which showed deep potential is very accurate compared to DFT results:
```python
from dpdata import System

model_predict = "./graph.000.pb"
system = System('./0.900/POSCAR')
predict_result=system.predict(model_predict)
```
![1699322148003](https://github.com/deepmodeling/deepmd-kit/assets/122458488/09b537d7-f08d-4fd2-a18e-0d023df4ac48)

Second is Lammps result, which showed my deep potential was far away from DFT:
```bash
variable        n               equal count(all)
variable        P               equal pe/$n
variable        V               equal vol

units           metal
boundary        p p p
atom_style      atomic

neighbor        1.0 bin
read_data       conf.lmp   

mass            1 132.905000  # Cs
mass            2 126.904000  # I
mass            3 207.200000  # Pb

pair_style      deepmd ../graph.000.pb 
pair_coeff      * *

thermo_style    custom step temp pe ke etotal press vol lx ly lz xy xz yz
thermo          10
dump            1 all custom 100 dump.xyz id type x y z fx fy fz

min_style       cg
minimize        1E-12 1E-12 100000 100000
```
![1699322357976](https://github.com/deepmodeling/deepmd-kit/assets/122458488/195113c7-a1f9-4e2b-83f9-9ae14676c436)

Then I noticed that dp.predict used the BFGS method in ASE package.
```python
    def minimize(
        self, *args: Any, minimizer: Union[str, Minimizer], **kwargs: Any
    ) -> "MultiSystems":
        """Minimize geometry by a minimizer.

        Parameters
        ----------
        *args : iterable
            Arguments passing to the minimizer
        minimizer : str or Minimizer
            The assigned minimizer
        **kwargs : dict
            Other arguments passing to the minimizer

        Returns
        -------
        MultiSystems
            A new labeled MultiSystems.

        Examples
        --------
        Minimize a system using ASE BFGS along with a DP driver:

        >>> from dpdata.driver import Driver
        >>> from ase.optimize import BFGS
        >>> driver = Driver.get_driver("dp")("some_model.pb")
        >>> some_system.minimize(minimizer="ase", driver=driver, optimizer=BFGS, fmax=1e-5)
        """
        if not isinstance(minimizer, Minimizer):
            minimizer = Minimizer.get_minimizer(minimizer)(*args, **kwargs)
        new_multisystems = dpdata.MultiSystems(type_map=self.atom_names)
        for ss in self:
            new_multisystems.append(ss.minimize(*args, minimizer=minimizer, **kwargs))
        return new_multisystems
```
I want to know that where this great difference come from? Is my potential is not accurate or just the optimization algorithm difference between BFGS and CG?

**Comment by PolyuWeldingSpock** [comment]
Since Lammps.minimize might give unreasonable result because of the unphysical configurations. I tried lammps with setting `run 0` 
```bash
units           metal
boundary        p p p
atom_style      atomic

neighbor        1.0 bin

read_data       conf.lmp   

mass            1 132.905000  # Cs
mass            3 207.200000  # Pb
mass            2 126.904000  # I
pair_style      deepmd ../graph.000.pb 
pair_coeff      * *

thermo_style    custom step temp pe ke etotal press vol lx ly lz xy xz yz
thermo          10
dump            1 all custom 100 dump.xyz id type x y z fx fy fz


thermo 5
thermo_style custom step pe ke etotal temp press vol
#velocity all create 300.0 789432 rot yes dist gaussian
fix 1 all nvt temp 300.0 300.0 ${Tdt} 
timestep ${dt}
run  0
```
The DP curve has similar trend with DFT. Although the values are still not correct.
![602adb69e9499c50d42774c6a2fb0b7d](https://github.com/deepmodeling/deepmd-kit/assets/122458488/33c8aabc-4bc4-4e91-976a-68ed2091aa04)

**Comment by njzjz** [comment]
> Then I noticed that dp.predict used the BFGS method in ASE package.

To minimize, `minimize` should be used instead of `predict`.

**Reply by PolyuWeldingSpock** [reply]
Thanks for your reply. I want to know why dp.predict give the great fitting with DFT results but lammps gives different values? Does dp.predict use dp driver or lammps driver to calculate the energy, forces and virials?

**Reply by njzjz** [reply]
Dpdata utilizes the Python interface. Dpdata and LAMMPS should have the same potential energy for the same structure. If not, the coordinates and cells are probably different.

**Reply by PolyuWeldingSpock** [reply]
Thanks for your reply!
